### PR TITLE
User Authntication: Add user authentication base on django authentication

### PR DESCRIPTION
**Pull request by ronyyosef**
Add login, signup, and logout functionality
image
image
#106 #99 #105 #104

### CODE REVIEW

File: accounts/tests/tests.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from django.template.loader import get_template
+
+from house.constants import USER_LOGIN_PAGE_ROUTE, USER_SIGNUP_ROUTE
+
+
+class TestLoginPage:
+    def test_user_login_template(self):
+        get_template(USER_LOGIN_PAGE_ROUTE)
```
Comment (LiorKGOW) [inline accounts/tests/tests.py:8]: Doesn't this function miss assert?

File: accounts/tests/tests.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    def test_login(self, client):
+        response = client.get('/accounts/login/')
+        assert response.status_code == 200
+        assert 'registration/login.html' in response.template_name
+
+
+class TestSignupPage:
+    def test_signup_template(self):
+        get_template(USER_SIGNUP_ROUTE)
```
Comment (LiorKGOW) [inline accounts/tests/tests.py:18]: Doesn't this function miss assert?

File: accounts/tests/tests.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from django.template.loader import get_template
+
+from house.constants import USER_LOGIN_PAGE_ROUTE, USER_SIGNUP_ROUTE
+
+
```
Comment (LiorKGOW) [inline accounts/tests/tests.py:5]: Are these tests alligned with what Barak told us our tests were missing (in the Demo)?
about the templates, that we need to check if the right template is presented

**Review by LiorKGOW — CHANGES_REQUESTED** [review]
Nice work @ronyyosef 
I have left you some comments 👍

File: Xpense/urls.py [hunk 4]
```diff
@@ -18,6 +18,8 @@
 
 urlpatterns = [
```
Comment (ShellyGolden) [inline Xpense/urls.py:19]: Where is the url for the house?
It canceled because the user?

File: accounts/tests/tests.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+from django.template.loader import get_template
+
```
Comment (ShellyGolden) [inline accounts/tests/tests.py:2]: unnecessary empty line

File: accounts/urls.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+from django.urls import path
+
+from .views import SignUpView
+
+urlpatterns = [
+    path("signup/", SignUpView.as_view(), name="signup"),
```
Comment (ShellyGolden) [inline accounts/urls.py:6]: What about url for the logout?

File: Xpense/settings.py [hunk 7]
```diff
@@ -125,3 +127,6 @@
 # https://docs.djangoproject.com/en/4.0/ref/settings/#default-auto-field
 
 DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'
+
+LOGIN_REDIRECT_URL = "/../house/"
+LOGOUT_REDIRECT_URL = "/"
```
[the remark this answers is not in the document]
Comment (ronyyosef) [reply Xpense/settings.py:132]: This means that after logout you will be redirect to "/"

File: accounts/tests/tests.py (hunk 1 above)
Comment (ronyyosef) [reply accounts/tests/tests.py:8]: Nope, If the get template will fail the test will fail with a message of the Exception

File: accounts/tests/tests.py (hunk 2 above)
Comment (ronyyosef) [reply accounts/tests/tests.py:18]: Same : Nope, If the get template will fail the test will fail with a message of the Exception

File: accounts/tests/tests.py (hunk 3 above)
Comment (ronyyosef) [reply accounts/tests/tests.py:5]: I think that it is.
@adiabramovitch what do you think about it?

File: Xpense/urls.py (hunk 4 above)
Comment (ronyyosef) [reply Xpense/urls.py:19]: its the last one

File: accounts/tests/tests.py (hunk 5 above)
Comment (ronyyosef) [reply accounts/tests/tests.py:2]: Why is it unnecessary? 2 lines between imports and function/class is required

File: accounts/urls.py (hunk 6 above)
Comment (ronyyosef) [reply accounts/urls.py:6]: It is available from this : 
path('accounts/', include("django.contrib.auth.urls"))
In xpense.urls

It is available at "accounts/logout"

File: Xpense/urls.py (hunk 4 above)
Comment (ShellyGolden) [reply Xpense/urls.py:19]: I missed it 
you can mark this conflict as resolve :)

File: accounts/tests/tests.py (hunk 5 above)
Comment (ShellyGolden) [reply accounts/tests/tests.py:2]: image
line 2

File: accounts/urls.py (hunk 6 above)
Comment (ShellyGolden) [reply accounts/urls.py:6]: thanks
you can mark this conflict as resolved :)

File: accounts/tests/tests.py (hunk 5 above)
Comment (ronyyosef) [reply accounts/tests/tests.py:2]: hoo, that because it two different types of imports, one import from the project files and one from 3rd party library

Comment (ShellyGolden) [reply accounts/tests/tests.py:2]: ok

[the remark this answers is not in the document]
Comment (LiorKGOW) [reply accounts/tests/tests.py:2]: To all the tests, maybe add some fixtures if needed for some

File: accounts/tests/tests.py (hunk 1 above)
Comment (LiorKGOW) [reply accounts/tests/tests.py:8]: Will it show up on the test description when you run pytest?

File: Xpense/urls.py [hunk 8]
```diff
@@ -18,6 +18,8 @@
...
     path('admin/', admin.site.urls),
+    path('accounts/', include("accounts.urls"), name="accounts"),
+    path('accounts/', include("django.contrib.auth.urls")),
```
[the remark this answers is not in the document]
Comment (LiorKGOW) [reply Xpense/urls.py:22]: ok 👍

**Review by LiorKGOW — APPROVED** [review]
I approve 👍 
GJ Rony !
I suggest waiting for a response from @adiabramovitch on the question about Barak's review for our tests

File: accounts/tests/tests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ronyyosef) [reply accounts/tests/tests.py:2]: I don't see here a reason for fixture or django_db
The test will run

**Review by ShellyGolden — COMMENTED** [review]
good job :)

File: accounts/tests/tests.py (hunk 3 above)
Comment (adiabramovitch) [reply accounts/tests/tests.py:5]: You are using the right object but you should check also the response URL to see that you have rendered the right template. 
for example: 
https://github.com/redhat-beyond/ResuMe/blob/08fc56c240518185ba5136f9f4583e01e538f8e0/users/test_users_views.py#L48
Also, I would check the object **response** to see which other information it holds so I can verify more information about the request. 

Also, you should check **bad login** and **bad signup** doesn't matter it is a Django framework as Barak explained last week/

File: templates/registration/login.html [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+        <h1>Login</h1>
+        <div class="card-body">
+            <form method="POST">
+                {% csrf_token %}
+                <fieldset class="form-group">
+                    {{ form|crispy }}
+                </fieldset>
+                <div class="form-group">
+                    <button class="btn btn-primary mt-3" type="submit">Log in</button>
+                    <a class="btn btn-primary mt-3" href="{% url 'signup' %}">Signup</a>
```
Comment (adiabramovitch) [inline templates/registration/login.html:16]: Sign Up or SIGN UP

File: templates/registration/signup.html [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+{% block content %}
+    <div class="card" style="color: rgb(0, 0, 0); background: rgba(252, 252, 252, 0.3); width: 500px;">
+        <h2>Sign up</h2>
+        <form method="post">
+            {% csrf_token %}
+            {{ form.as_p }}
+            <div class="form-group">
+                <button class="btn btn-primary mt-3" type="submit">Sign Up</button>
+                <a class="btn btn-primary mt-3" href="{% url 'login' %}">Log in</a>
+            </div>
```
Comment (adiabramovitch) [inline templates/registration/signup.html:11]: Login 
it is not consistent with templates/registration/login.html

File: templates/registration/signup.html [hunk 11]
```diff
@@ -0,0 +1,14 @@
+{% extends "base.html" %} {% load static %}
+{% block content %}
+    <div class="card" style="color: rgb(0, 0, 0); background: rgba(252, 252, 252, 0.3); width: 500px;">
+        <h2>Sign up</h2>
+        <form method="post">
+            {% csrf_token %}
+            {{ form.as_p }}
+            <div class="form-group">
+                <button class="btn btn-primary mt-3" type="submit">Sign Up</button>
```
Comment (adiabramovitch) [inline templates/registration/signup.html:9]: it is not consistent to templates/registration/login.html

**Review by adiabramovitch — CHANGES_REQUESTED** [review]
Can you place the forms in the center of the page?
image

**Comment by ronyyosef** [comment]
> Can you place the forms in the center of the page? image

We will open another PR to make the UI look nicer

**Comment by ronyyosef** [comment]
image
image

**Comment by ronyyosef** [comment]
Updated tests and sign up sign in view

File: static/css/accounts/signup.css [hunk 12]
```diff
@@ -0,0 +1,5 @@
+#signup{
+    align-content: center;
+    margin-top: 5%;
+    margin-bottom: auto;
+}
```
Comment (adiabramovitch) [inline static/css/accounts/signup.css:5]: missing new line

File: static/css/accounts/login.css [hunk 13]
```diff
@@ -0,0 +1,5 @@
+#login{
+    align-content: center;
+    margin-top: 5%;
+    margin-bottom: auto;
+}
```
Comment (adiabramovitch) [inline static/css/accounts/login.css:5]: missing new line

**Review by adiabramovitch — COMMENTED** [review]
Are you adding it in the second PR? 
image

And What about 400's code tests?

**Comment by ronyyosef** [comment]
> Are you adding it in the second PR? image
> 
> And What about 400's status code tests?

yes the "after login" will be add at the next pr
What tests do you mean, where should i get a 400?
django returning 200 with the errors when there is error in the user input

File: accounts/tests/tests.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    def test_login_user(self, client, new_user):
+        response = client.post('/accounts/login/', {'username': USERNAME, 'password': PASSWORD})
+        assert response.status_code == 302
+        assert response.url == '/../house/'
+
+    def test_bad_login(self, client):
+        response = client.post('/accounts/login/', {'username': USERNAME, 'password': PASSWORD})
+        assert response.status_code == 200
+        assert 'registration/login.html' in response.template_name
```
Comment (adiabramovitch) [inline accounts/tests/tests.py:41]: How is this a bad login? it seems like the previous test

Comment (ronyyosef) [reply accounts/tests/tests.py:41]: User never created so the login will fail

**Pull request merged**